### PR TITLE
Remove unused splitByTrace from groupbytraceprocessor, add tests to batchpersignal

### LINKED ISSUE #3090: Loki exporter : Lot of out of order errors
**Describe the bug**
When a use Loki exporter with a unique set of labels I get a lot of out of order dropped messages (resulting in a full OTEL collector queue).
My labels are correct accross logs to be uniques by main producers processes (but I can't seriously introduce thread ID as label).

Is there a workarond for now ?

**Steps to reproduce**
Create an app which sends logs through OTLP to a collector.
Different threads of the app emits logs, so timestamps can be a little shuffled when OTEL Collector receive them.

**What did you expect to see?**
Especialy for Loki exporter (or as a new processor), we should be able to buffer logs for a short time and order them before send them to Loki.

**What did you see instead?**
OTEL Collector queue is filling a lot with retries for loki 400 bad resquests responses (out of order).

**What version did you use?**
otel-collector-contrib:523e417c06215ffdd5ce1d36ad4fa657f4dd525c

**What config did you use?**
```yaml
receivers:
  otlp:
    protocols:
      http:
      grpc:

exporters:
  logging:
    loglevel: debug
  loki:
    endpoint: http://localhost:3100/loki/api/v1/push
    labels:
      attributes:
        [... mappings for otel attributes...]

processors:
  batch:

extensions:
  health_check:

service:
  extensions: [health_check]
  pipelines:
    logs:
      receivers: [otlp]
      processors: [batch]
      exporters: [logging, loki]
    traces:
      receivers: [otlp]
      processors: [batch]
      exporters: [logging]
    metrics:
      receivers: [otlp]
      processors: [batch]
      exporters: [logging]
```

**Additional context**
Could be fixed by https://github.com/grafana/loki/issues/1544

### PR DESCRIPTION
**Pull request by LouisStAmour**
I was looking at groupbytraceprocessor to try and build a log processor that would collect an ordered ring buffer (circular queue) of logs using insertion sort and emit a sorted part of the buffer after some delay to fix #3090. This PR doesn't do that - while investigating the rather complicated (to me) control flow of groupbytraceprocessor, I discovered that the splitByTrace function it had was only referenced in tests. So this PR removes that function. In doing so, I ported existing tests from groupbytraceprocessor to batchpersignal and discovered that how Nil was handled changed when the batchpersignal function was added to groupbytraceprocessor.

As noted in the commit text: Note that previously groupbytraceprocessor expected Nil trace IDs to be removed, now they are kept. This might have further consequences for groupbytraceprocessor, I couldn’t find any tests for it that included Nil trace IDs. That said, the only consequence I can think of is that it would attempt to group all Nil trace IDs together, which is possibly a bug. I’d imagine the better behaviour would be to emit/pass through any traces with Nil trace IDs immediately? But the switch to using batchpersignal has already occurred, so fixing this is out of scope for this commit.

### CODE REVIEW

File: processor/groupbytraceprocessor/processor.go [hunk 1]
```diff
@@ -245,58 +245,4 @@ func (sp *groupByTraceProcessor) onTraceRemoved(traceID pdata.TraceID) error {
...
-			}
-
-			// there is only one instrumentation library per batch
-			batches[sTraceID].rs.InstrumentationLibrarySpans().At(0).Spans().Append(span)
-		}
-	}
-
-	return result
-}
+}
```
Comment (bogdandrutu) [inline processor/groupbytraceprocessor/processor.go:248]: New line.

File: pkg/batchpersignal/batchpersignal_test.go [hunk 2]
```diff
@@ -95,3 +157,65 @@ func TestSplitDifferentLogsIntoDifferentBatches(t *testing.T) {
...
+	assert.Equal(t, pdata.NewTraceID([16]byte{1, 2, 3, 4}), batches[0].ResourceLogs().At(0).InstrumentationLibraryLogs().At(0).Logs().At(0).TraceID())
+	assert.Equal(t, firstLibrary.Name(), batches[0].ResourceLogs().At(0).InstrumentationLibraryLogs().At(0).InstrumentationLibrary().Name())
+	assert.Equal(t, firstLog.Name(), batches[0].ResourceLogs().At(0).InstrumentationLibraryLogs().At(0).Logs().At(0).Name())
+	assert.Equal(t, secondLog.Name(), batches[0].ResourceLogs().At(0).InstrumentationLibraryLogs().At(0).Logs().At(1).Name())
+
+	// second batch
+	assert.Equal(t, pdata.NewTraceID([16]byte{1, 2, 3, 4}), batches[1].ResourceLogs().At(0).InstrumentationLibraryLogs().At(0).Logs().At(0).TraceID())
+	assert.Equal(t, secondLibrary.Name(), batches[1].ResourceLogs().At(0).InstrumentationLibraryLogs().At(0).InstrumentationLibrary().Name())
+	assert.Equal(t, thirdLog.Name(), batches[1].ResourceLogs().At(0).InstrumentationLibraryLogs().At(0).Logs().At(0).Name())
+}
```
Comment (bogdandrutu) [inline pkg/batchpersignal/batchpersignal_test.go:221]: New line.

**Review by bogdandrutu — COMMENTED** [review]
@jpkrohling can you also take a look?

**Comment by bogdandrutu** [comment]
Please also fix lint issues:
```
make[1]: Entering directory '/home/circleci/project/pkg/batchpertrace'
golangci-lint run --allow-parallel-runners
batchpersignal_test.go:221: File is not `gofmt`-ed with `-s` (gofmt)
}
make[1]: *** [../../Makefile.Common:93: lint] Error 1
```

**Comment by jpkrohling** [comment]
>while investigating the rather complicated (to me) control flow of groupbytraceprocessor

Indeed, it's now more complicated than it deserves. The original idea was to simplify the concurrency access to the underlying data structures, but I would probably do it differently today. 

>I discovered that the splitByTrace function it had was only referenced in tests

Good catch, this was probably a leftover of the change that moved this logic to the batch per signal library.

>discovered that how Nil was handled changed when the batchpersignal function was added to groupbytraceprocessor.

Batch per signal was introduced in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/2549, to make the load balancing exporter work with logs as well. I believe this change might have been originated there. Either that or when the traceID changed from being a byte array to its own structure.

**Review by jpkrohling — APPROVED** [review]

**Comment by tigrannajaryan** [comment]
The linting failed and this seems stale now. Is there an intent to continue working on this PR?

**Comment by LouisStAmour** [comment]
Happy to fix the PR, I've added it to my todo list for tomorrow. Sorry, I'm a bit new to more formal open source contributions. :)

**Comment by jpkrohling** [comment]
This needs a rebase now.

**Comment by LouisStAmour** [comment]
My first force-push was a rebase against origin/main, the second fixed a gofmt error in one file. I couldn't find any lint errors or gofmt errors outside of that. Also ran gotidy and didn't notice any issues against processor/groupbytraceprocessor.

**Comment by bogdandrutu** [comment]
Please fix the tests

**Comment by LouisStAmour** [comment]
I've rebased against main in the hopes that tests will fix themselves. They seemed to fail due to a recent commit for oauth and some windows test failures.

**Review by jpkrohling — APPROVED** [review]

**Pull request merged**